README.md

```diff
@@ -1 +1,218 @@
+
+What should we do?
+
+## Async Error Handling
+
+Node developers use error-as-the-first-argument pattern to bubble up the errors. When you have an error which comes from an asynchronous code, check for it. Always check for these errors and pass them up the chain. 
+
+Consider this code in which we have a `callback` function and use it to pass either `null` or `error`:
+
+```js
+module.exports = function(inputFile, customers, callback){
+  return function writeFile(error){
+    if (!error) {
+      return callback(null, customers)
+    } else {
+      return callback(error)
+    }
+  }
+}
+```
+
+The same code can be refactored into more compact version:
+
+```js
+module.exports = function(inputFile, customers, callback){
+  return function writeFile(error){
+    if (error) return callback(error)
+    return callback(null, customers)
+  }
+}
+```
+
+By using `if (error) return callback(error)` you will NOT increase the indentation which leads to callback hell. As you know, `return` will stop the execution and return the value making this function an expression. We don't really care about it being the expression. We use `return` to control the flow.
+
+If you don't have an error (it's falsy), just continue with your logic. At some point, you might want to return some data back to the original caller via the callback. Pass the data as the second argument, leaving the first for the error which is `null` if everything went fine.
+
+Because of the way closures work in JavaScript/Node (they have access to outer scope), you can invoke `callback` at any level when you have multiple levels of nested callbacks:
+
+```js
+var fs = require('fs')
+var inputFile = './customers.csv'
+
+module.exports = function(callback){
+  var customers = []
+  fs.stat(inputFile, function(error, stats){
+    if (stats.isFile()) {
+      var customersCSV = fs.readFile(inputFile, 'utf8', function(error, customersCSV){
+        if (!error) {
+          ...
+          fs.writeFile('./customers.json', customersJSON, 'utf-8', function(error){
+            if (!error) {
+              return callback(null, customers)
+            } else {
+              return callback(error)
+            }
+          })
+        } else {
+          return callback(error)
+        }
+      })
+    } else {
+      console.error('The input CSV file is not found')
+    }
+  })
+}
+```
+
+When using an object which implements Event Emitters, listen to `error` event with `on('error', callback)`. Similarly, when you implement your own object with events, trigger error with `emit('error', error)`. As you remember, you can pass any data with `emit()`.
+
+When all hell breaks lose, listen to `process.on('uncaughtException', callback)`. Think about it as the last resort. Don't try to resume the execution at this point. There might be some lost states. Just log and exit gracefully. 
```

Comment:
I think students could benefit from seeing an example of this in lieu of the example above (they'll already have a reasonably solid grasp of scope in JS)